File: src/main/java/core/db/ConnectionManager.java
package core.db;

import core.util.ExceptionUtils;
import core.util.HOLogger;

import java.sql.*;

/**
 * Provides the connection functions to the database
 */
public class ConnectionManager {
	Connection connection;
	private Statement statement;
	private DBInfo dbInfo;

	private StatementCache statementCache;

	/**
	 * Closes the connection
	 */
	public final void disconnect() {
		try {
			var statement = connection.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
			statement.execute("SHUTDOWN");
			connection.close();
			connection = null;
		} catch (Exception e) {
			HOLogger.instance().error(getClass(), "JDBCAdapter.disconnect : " + e);
			connection = null;
		}
	}

	/**
	 * Execute a SQL Select statement
	 * 
	 * @param sqlStatement
	 *            Sql query with placeholders
	 * 
	 * @return ResultSet of the query
	 */
	public final ResultSet executeQuery(String sqlStatement) {
		try {
			if (connection.isClosed()) {
				return null;
			}
			return statement.executeQuery(sqlStatement);
		} catch (Exception e) {
			HOLogger.instance().error(
					getClass(),
					"executeQuery : " + e + "\nStatement: " + sqlStatement + "\n"
							+ ExceptionUtils.getStackTrace(e));
		}
		return null;
	}

	public final PreparedStatement createPreparedStatement(String sql) {
		return statementCache.getPreparedStatement(sql);
	}

	public final ResultSet executePreparedQuery(String query, Object... params) {
		return executePreparedQuery(statementCache.getPreparedStatement(query), params);
    }

	private ResultSet executePreparedQuery(PreparedStatement preparedStatement, Object... params) {
		try {
			if (connection.isClosed()) {
				return null;
			}
			int i = 0;
			for (var p : params) {
				preparedStatement.setObject(++i, p);
			}
			return preparedStatement.executeQuery();
		} catch (Exception e) {
			HOLogger.instance().error(getClass(), "executePreparedQuery : " + e
					+ "\nStatement: " + preparedStatement
					+ "\n" + ExceptionUtils.getStackTrace(e));
			return null;
		}
	}

	/**
	 * Executes an SQL INSERT, UPDATE or DELETE statement. In addition, SQL
	 * statements that return nothing, such as SQL DDL statements, can be
	 * executed.
	 * 
	 * @param sqlStatement
	 *            INSERT, UPDATE or DELETE statement
	 * 
	 * @return either the row count for SQL Data Manipulation Language (DML)
	 *         statements or 0 for SQL statements that return nothing
	 * 
	 */
	public final int executeUpdate(String sqlStatement) {
		int ret = 0;

		try {
			if (connection.isClosed()) {
				return 0;
			}
			// HOLogger.instance().log(getClass(), Sql );
			ret = statement.executeUpdate(sqlStatement);
			return ret;
		} catch (Exception e) {
			HOLogger.instance().error(
					getClass(),
					"JDBCAdapter.executeUpdate : " + e + "\nStatement: " + sqlStatement + "\n"
							+ ExceptionUtils.getStackTrace(e));
			return 0;
		}
	}

	public final int executePreparedUpdate(String insert, Object... params) {
		return executePreparedUpdate(statementCache.getPreparedStatement(insert), params);
	}

	private int executePreparedUpdate(PreparedStatement preparedStatement, Object... params) {
		int ret;

		try {
			if (connection.isClosed()) {
				return 0;
			}
			int i = 0;
			for ( var p: params) {
				preparedStatement.setObject(++i, p);
			}
			ret = preparedStatement.executeUpdate();
			return ret;
		} catch (Exception e) {
			HOLogger.instance().error(
					getClass(),
					"JDBCAdapter.executeUpdate : " + e + "\nStatement: " + preparedStatement.toString() + "\n"
							+ ExceptionUtils.getStackTrace(e));
			return 0;
		}
	}

	/**
	 * Connects to the requested database
	 * 
	 * @param URL
	 *            The path to the Server
	 * @param User
	 *            User
	 * @param PWD
	 *            Password
	 * @param Driver
	 *            The driver to user
	 * 
	 */
	public final void connect(String URL, String User, String PWD, String Driver) throws Exception {
		// Initialise the Database Driver Object
		Class.forName(Driver);
		var connection = DriverManager.getConnection(URL, User, PWD);
		connect(connection);
	}

	public final void connect(Connection conn) throws Exception {
		try {
			connection = conn;
			statement = connection.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
			statementCache = new StatementCache(this);
		} catch (SQLException e) {
			if (connection != null) {
				try {
					connection.close();
				} catch (Exception ex) {
					HOLogger.instance().error(getClass(), "ConnectionManager.connect : " + ex.getMessage());
				}
			}
			HOLogger.instance().error(getClass(), "ConnectionManager.connect : " + e.getMessage());
			throw e;
		}
	}

	/**
	 * 
	 * @return DBInfo
	 * @throws Exception
	 */
	public DBInfo getDBInfo() throws Exception {
		if (dbInfo == null)
			dbInfo = new DBInfo(connection.getMetaData());
		return dbInfo;
	}

	public Object[] getAllTableNames() {
		try {
			return getDBInfo().getAllTablesNames();
		} catch (Exception e) {
			HOLogger.instance().error(getClass(), "JDBCAdapter.getAllTableNames : " + e);
			return new String[] { e.getMessage() };
		}
	}
}

File: src/main/java/core/db/ConnectionManager.kt
package core.db

import core.util.HOLogger

import java.sql.*

/**
 * Provides the connection functions to the database
 */
class ConnectionManager {
    var connection: Connection? = null

    var statement: Statement? = null

    var dbInfo: DBInfo? = null
        get() {
            if (field == null) {
                field = DBInfo(connection?.metaData)
            }
            return field
        }

    var statementCache: StatementCache = StatementCache(this)

    /**
     * Closes the connection
     */
    fun disconnect() {
        try {
            val statement = connection?.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY)
            statement?.execute("SHUTDOWN")
            connection?.close()
        } catch (e: Exception) {
            HOLogger.instance().error(javaClass, "ConnectionManager.disconnect : $e")
        } finally {
            connection = null
        }
    }

    /**
     * Execute a SQL Select statement
     *
     * @param sqlStatement
     *            Sql query with placeholders
     *
     * @return ResultSet of the query
     */
    fun executeQuery(sqlStatement: String): ResultSet? {
        checkConnectionNotClosed()
        return statement?.executeQuery(sqlStatement)

    }

    private fun checkConnectionNotClosed() {
        if (connection?.isClosed == true) {
            throw SQLException("Connection closed")
        }
    }

    fun executePreparedQuery(query: String, vararg params: Any): ResultSet? {
        val preparedStatement = statementCache.getPreparedStatement(query)
        if (preparedStatement != null) {
            return executePreparedQuery(preparedStatement, *params)
        }
        return null
    }

    private fun executePreparedQuery(preparedStatement: PreparedStatement, vararg params: Any): ResultSet {
        checkConnectionNotClosed()
        for ((i, p) in params.withIndex()) {
            preparedStatement.setObject(i + 1, p)
        }
        return preparedStatement.executeQuery()
    }

    /**
     * Executes an SQL INSERT, UPDATE or DELETE statement. In addition, SQL
     * statements that return nothing, such as SQL DDL statements, can be
     * executed.
     *
     * @param sqlStatement
     *            INSERT, UPDATE or DELETE statement
     *
     * @return either the row count for SQL Data Manipulation Language (DML)
     *         statements or 0 for SQL statements that return nothing
     *
     */
    fun executeUpdate(sqlStatement: String): Int {
        checkConnectionNotClosed()
        return statement?.executeUpdate(sqlStatement) ?: 0
    }

    fun executePreparedUpdate(insert: String, vararg params: Any): Int {
        val preparedStatement = statementCache.getPreparedStatement(insert)
        if (preparedStatement != null) {
            return executePreparedUpdate(preparedStatement, *params)
        }
        return 0
    }

    private fun executePreparedUpdate(preparedStatement: PreparedStatement, vararg params: Any): Int {
        for ((i, p) in params.withIndex()) {
            preparedStatement.setObject(i + 1, p)
        }
        return preparedStatement.executeUpdate()
    }

    /**
     * Connects to the requested database
     *
     * @param url
     *            The path to the Server
     * @param user
     *            User
     * @param password
     *            Password
     * @param driver
     *            The driver to user
     *
     */
    fun connect(url: String, user: String, password: String, driver: String) {
        // Initialise the Database Driver Object
        Class.forName(driver)
        val connection = DriverManager.getConnection(url, user, password)
        connect(connection)
    }

    fun connect(conn: Connection) {
        connection = conn
        statement = conn.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY)
        statementCache = StatementCache(this)
    }

    fun getAllTableNames(): Array<String> {
        try {
            return dbInfo?.getAllTablesNames() ?: emptyArray()
        } catch (e: Exception) {
            HOLogger.instance().error(javaClass, "ConnectionManager.getAllTableNames : $e")
            throw e
        }
    }
}
